contracts/pallet-ibc/src/ics20_fee/mod.rs

```diff
@@ -0,0 +1,281 @@
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		order: Order,
+		connection_hops: &[ConnectionId],
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		counterparty: &Counterparty,
+		version: &Version,
+		counterparty_version: &Version,
+		relayer: &Signer,
+	) -> Result<Version, Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_open_try(
+			&mut ctx,
+			output,
+			order,
+			connection_hops,
+			port_id,
+			channel_id,
+			counterparty,
+			version,
+			counterparty_version,
+			relayer,
+		)
+	}
+
+	fn on_chan_open_ack(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		counterparty_version: &Version,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_open_ack(
+			&mut ctx,
+			output,
+			port_id,
+			channel_id,
+			counterparty_version,
+			relayer,
+		)
+	}
+
+	fn on_chan_open_confirm(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_open_confirm(&mut ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_chan_close_init(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_close_init(&mut ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_chan_close_confirm(
+		&mut self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		port_id: &PortId,
+		channel_id: &ChannelId,
+		relayer: &Signer,
+	) -> Result<(), Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		self.inner.on_chan_close_confirm(&mut ctx, output, port_id, channel_id, relayer)
+	}
+
+	fn on_recv_packet(
+		&self,
+		_ctx: &dyn ModuleCallbackContext,
+		output: &mut ModuleOutputBuilder,
+		packet: &Packet,
+		relayer: &Signer,
+	) -> Result<Acknowledgement, Ics04Error> {
+		let mut ctx = Context::<T>::default();
+		let packet_data =
+			serde_json::from_slice::<PacketData>(packet.data.as_slice()).map_err(|e| {
+				Ics04Error::implementation_specific(format!("Failed to decode packet data {:?}", e))
+			})?;
+		let percent = ServiceCharge::<T>::get().unwrap_or(T::ServiceCharge::get());
+		let fee = percent * packet_data.token.amount.as_u256().as_u128();
+		// Send full amount to receiver
+		let ack = self.inner.on_recv_packet(&mut ctx, output, packet, relayer)?;
+		// Send fee from the receiver's account to the pallet account
+		let receiver = <T as crate::Config>::AccountIdConversion::try_from(packet_data.receiver)
+			.map_err(|_| Ics04Error::implementation_specific("Failed to parse account".to_string()))
+			.expect(
```

Comment:
I'm not seeing where it has been validated. Even so, let's return a proper error, or convert packet_data's in a different type during the validation step so that downstream we don't need further unwraps.

Reply:
If the `self.inner.on_recv_packet`  called before this does not fail then it must have done the account parsing successfully.

Reply:
Yeah, but that is not checked by the compiler, so if later other devs work on this codebase, this can cause issues. I think you can resolve this by:

1. Changing the trait of Module to something like this:

```
trait Module {
    // This is the go style function, which uses a string to encode account ids since Go's type system sucks
    fn foo(s: String) -> Result<T> {
       let account = s.try_into()?;
       Ok(Self::bar(account))
    }
    
    // This is what Rust modules should prefer to implement, because it is more typed. There is a default implementation 
    // of foo if bar is implemented.
    fn bar(s: AccountId) -> T;
}
```

This way, we can leverage the type system to proof that validation has already occured.

Otherwise I would just return an error here.